testpb/proto_methods_test.go

```diff
@@ -0,0 +1,204 @@
+package testpb
+
+import (
+	"github.com/stretchr/testify/require"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/runtime/protoiface"
+	"google.golang.org/protobuf/runtime/protoimpl"
+	"pgregory.net/rapid"
+	"testing"
+)
+
+func TestProtoMethods(t *testing.T) {
+	t.Run("testSize", rapid.MakeCheck(testSize))
+	t.Run("testMarshal", rapid.MakeCheck(testMarshal))
+	t.Run("testUnmarshal", rapid.MakeCheck(testUnmarshal))
+}
+
+func testSize(t *rapid.T) {
+	slowMsg := getRapidMsg(t)
+	fastMsg := slowMsg.ProtoReflect()
+
+	methods := fastMsg.ProtoMethods()
+
+	result := methods.Size(protoiface.SizeInput{Message: fastMsg})
+	expected := proto.Size(fastMsg.Interface())
+
+	require.Equal(t, expected, result.Size)
+}
+
+func testMarshal(t *rapid.T) {
+	slowMsg := getRapidMsg(t)
+	fastMsg := slowMsg.ProtoReflect()
+
+	methods := fastMsg.ProtoMethods()
+
+	result, err := methods.Marshal(protoiface.MarshalInput{
+		NoUnkeyedLiterals: struct{}{},
+		Message:           fastMsg,
+		Buf:               nil,
+		Flags:             0,
+	})
+	require.NoError(t, err)
+
+	canonical, err := proto.MarshalOptions{Deterministic: true}.Marshal(fastMsg.Interface())
+	require.NoError(t, err)
+
+	require.Equal(t, canonical, result.Buf)
+}
+
+func testUnmarshal(t *rapid.T) {
+	slowMsg := getRapidMsg(t)
+	fastMsg := slowMsg.ProtoReflect()
+
+	proto.Marshal(fastMsg.Interface())
+
+	methods := fastMsg.ProtoMethods()
+
+	testCases := []struct {
+		name      string
+		unmarshal func(bz []byte, msg *A)
+		marshal   func(msg proto.Message) []byte
+	}{
+		{
+			name: "proto.Marshal pulsar Unmarshal",
+			marshal: func(msg proto.Message) []byte {
+				bz, err := proto.Marshal(msg)
+				require.NoError(t, err)
+				return bz
+			},
+			unmarshal: func(bz []byte, msg *A) {
+				_, err := methods.Unmarshal(protoiface.UnmarshalInput{
+					NoUnkeyedLiterals: struct{}{},
+					Message:           msg.ProtoReflect(),
+					Buf:               bz,
+					Flags:             0,
+					Resolver:          nil,
+				})
+				require.NoError(t, err)
+			},
+		},
+		{
+			name: "pulsar.Marshal, proto.Unmarshal",
+			marshal: func(msg proto.Message) []byte {
+				res, err := methods.Marshal(protoiface.MarshalInput{
+					NoUnkeyedLiterals: struct{}{},
+					Message:           msg.ProtoReflect(),
+					Buf:               nil,
+					Flags:             0,
+				})
+				require.NoError(t, err)
+				return res.Buf
+			},
+			unmarshal: func(bz []byte, msg *A) {
+				err := proto.Unmarshal(bz, msg)
+				require.NoError(t, err)
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Logf("Test: %s\n", tc.name)
+		aa := A{}
+		bz := tc.marshal(fastMsg.Interface())
+		tc.unmarshal(bz, &aa)
+		fastAA := fastReflection_A(aa)
+		underlying := fastMsg.(*fastReflection_A)
+		CheckEqual(t, fastAA, *underlying)
+	}
+}
+
+func CheckEqual(t *rapid.T, a1, a2 fastReflection_A) {
```

Comment:
this function is kinda ugly, but other ways were comparing pointer addresses which were always different and in the unmarshaler case, the message unmarshaled with slices as `[]SliceType(nil)` which was throwing errors when checked against simply empty slice types